lib/puppet_x/templates/invoke_ps_command.erb

```diff
@@ -0,0 +1,165 @@
+$hostSource = @"
+using System;
+using System.Globalization;
+using System.Management.Automation.Host;
+using System.Threading;
+
+public class PuppetPSHost : PSHost
+{
+    private Guid _hostId = Guid.NewGuid();
+    private bool shouldExit;
+    private int exitCode;
+
+    public bool ShouldExit
+    {
+        get { return this.shouldExit; }
+    }
+
+    public int ExitCode
+    {
+        get { return this.exitCode; }
+    }
+
+    public void ResetExitStatus()
+    {
+      this.exitCode = 0;
+      this.shouldExit = false;
+    }
+
+    public override Guid InstanceId
+    {
+        get { return _hostId; }
+    }
+
+    public override string Name
+    {
+        get { return "PuppetPSHost"; }
+    }
+
+    public override Version Version
+    {
+        get { return new Version(1, 0); }
+    }
+
+    public override PSHostUserInterface UI
+    {
+        get { return null; }
+    }
+
+    public override CultureInfo CurrentCulture
+    {
+        get { return Thread.CurrentThread.CurrentCulture; }
+    }
+
+    public override CultureInfo CurrentUICulture
+    {
+        get { return Thread.CurrentThread.CurrentUICulture; }
+    }
+
+    public override void EnterNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void ExitNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void NotifyBeginApplication()
+    {
+        return;
+    }
+
+    public override void NotifyEndApplication()
+    {
+        return;
+    }
+
+    public override void SetShouldExit(int exitCode)
+    {
+      this.shouldExit = true;
+      this.exitCode = exitCode;
+    }
+}
+"@
+
+Add-Type -TypeDefinition $hostSource -Language CSharp
+
+$outputEvent = [Threading.EventWaitHandle]::OpenExisting("<%= output_ready_event_name %>")
+
+if ($runspace -eq $null)
+{
+  # CreateDefault2 requires PS3
+  if ([InitialSessionState].GetMethod('CreateDefault2'))
+  {
+    $sessionState = [InitialSessionState]::CreateDefault2()
+  }
+  else
+  {
+    $sessionState = [InitialSessionState]::CreateDefault()
+  }
+  $puppetPSHost = [PuppetPSHost]::new()
+  $runspace = [RunspaceFactory]::CreateRunspace($puppetPSHost, $sessionState)
+  $runspace.Open()
+}
+
+$powershell_code = @'
+<%= powershell_code %>
+'@
+
+try
+{
+  $ps = $null
+  $puppetPSHost.ResetExitStatus()
+  $runspace.ResetRunspaceState()
+
+  $ps = [PowerShell]::create()
+  $ps.Runspace = $runspace
+  [Void]$ps.AddScript($powershell_code)
+
+  $asyncResult = $ps.BeginInvoke()
+
+  if (!$asyncResult.AsyncWaitHandle.WaitOne(<%= timeout_ms %>, $false))
+  {
+    throw "Catastrophic failure: PowerShell resource timeout (<%= timeout_ms %> ms) exceeded while executing"
+  }
+
+  # doesn't generate output - just exits
+  $output = $ps.EndInvoke($asyncResult)
+
+  return @{
+    output       = $output
```

Comment:
Renamed to stdout.

I didn't do anything with `$Error` because its a hash of all the errors in the current session, with no differentiation between non-terminating and terminating errors. If it's a terminating error, it would be caught by our catch block and reported there. If it was non-terminating, we decided to continue not doing anything with them. So I didn't see the need for doing anything with `$Error`. Am I missing something?